Spigot-API-Patches/0199-Prioritise-own-classes-where-possible.patch

```diff
@@ -0,0 +1,132 @@
+From b111e02a4924fe7576f6edc2703ab8b656919c4b Mon Sep 17 00:00:00 2001
+From: Mariell Hoversholm <proximyst@proximyst.com>
+Date: Mon, 27 Apr 2020 12:31:59 +0200
+Subject: [PATCH] Prioritise own classes where possible
+
+This adds the plugin.yml property `classloader-prioritize-self` to allow
+plugins to decide that they want the classloader to prioritize the
+classes of the plugin it is loading for.
+
+This value is by default `false` and this will therefore not break any
+intended behaviour by default, but if a plugin knows it does not care
+about other plugins' classes, it can turn on the option to have it load
+its own classes in any case it can. This is especially useful with
+plugins using libraries which cannot have version incompatibilities
+and/or libraries which somehow break upon relocation.
+
+If a class is not found in the same jar as it is loading for and it does
+find it elsewhere, it will still choose the class elsewhere. This is
+intended behaviour, as it will only prioritise classes it has in its own
+jar, no other plugins' classes will be prioritised in any other order
+than the one they were registered in.
+
+diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+index 70d50819..418c0a03 100644
+--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
++++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+@@ -129,7 +129,11 @@ import org.yaml.snakeyaml.nodes.Tag;
+  *     <td><code>api-version</code></td>
+  *     <td>{@link #getAPIVersion()}</td>
+  *     <td>The API version which this plugin was programmed against</td>
+- * </tr>
++ * </tr><tr> <!-- Paper start -->
++ *     <td><code>classloader-prioritize-self</code></td>
++ *     <td>{@link #shouldClassLoaderPrioritizeOwnClasses()}</td>
++ *     <td>Whether the classloader should prioritize its own classes.</td>
++ * </tr> <!-- Paper end -->
+  * </table>
+  * <p>
+  * A plugin.yml example:<blockquote><pre>
+@@ -148,6 +152,8 @@ import org.yaml.snakeyaml.nodes.Tag;
+  *depend: [NewFire, FlameWire]
+  *api-version: 1.13
+  *
++ *classloader-prioritize-self: true<!-- Paper -->
```

Comment:
yaml doesn't use html comments does it?

Reply:
No, YAML definitely doesn't use HTML comment, it has a [Ruby-inspired comments syntax](https://yaml.org/spec/1.2/spec.html#id2780069): `#` after whitespace, so this should probably be

```suggestion
- *classloader-prioritize-self: true<!-- Paper -->
+ *classloader-prioritize-self: true # Paper
```

Reply:
That is actually a JavaDoc comment; it is not rendered in IntelliJ nor in actual JavaDocs (HTML), just as we don't use JD comments when we do paper start & end. If wanted, I'm happy to still change it out for the suggested, though, @aikar?

Reply:
Ah ok, leave as is then.